pkg/virt-api/rest/authorizer.go

```diff
@@ -123,82 +120,110 @@ func (a *authorizor) GetExtraPrefixHeaders() []string {
+	pathSplit := strings.Split(req.Request.URL.Path, "/")
+	if len(pathSplit) < 5 {
+		// When generateAccessReview is called isInfoOrHealthEndpoint already confirmed that len(pathSplit) >= 5,
+		// so this should never happen
+		return nil, fmt.Errorf("unknown api endpoint: %s", req.Request.URL.Path)
+	}
 
-	resourceAttributes, err := getResourceAttributes(pathSplit, httpRequest)
-	if err != nil {
-		return nil, err
+	// "namespaces" after version means that the URL points to a namespaced subresource
+	if pathSplit[4] == "namespaces" {
+		if err := addNamespacedAttributes(req, r, pathSplit); err != nil {
+			return nil, err
+		}
+	} else {
+		if err := addNonNamespacedAttributes(req, r, pathSplit); err != nil {
+			return nil, err
+		}
 	}
 
-	r.Spec.ResourceAttributes = resourceAttributes
 	return r, nil
 }
 
-func getResourceAttributes(pathSplit []string, httpRequest *http.Request) (*authorization.ResourceAttributes, error) {
+func addNamespacedAttributes(req *restful.Request, r *authv1.SubjectAccessReview, pathSplit []string) error {
 	// URL example
 	// /apis/subresources.kubevirt.io/v1alpha3/namespaces/default/virtualmachineinstances/testvmi/console
 	if len(pathSplit) < 9 {
-		return nil, fmt.Errorf("unknown api endpoint %s", httpRequest.URL.Path)
+		return fmt.Errorf("unknown api endpoint %s", req.Request.URL.Path)
 	}
 
-	group := pathSplit[2]
-	version := pathSplit[3]
-	namespace := pathSplit[5]
 	resource := pathSplit[6]
-	resourceName := pathSplit[7]
-	subresource := pathSplit[8]
-
 	if resource != "virtualmachineinstances" && resource != "virtualmachines" {
-		return nil, fmt.Errorf("unknown resource type %s", resource)
+		return fmt.Errorf("unknown resource type %s", resource)
 	}
 
-	verb, err := mapHttpVerbToRbacVerb(httpRequest.Method, resourceName)
+	resourceName := pathSplit[7]
+	verb, err := mapHttpVerbToRbacVerb(req.Request.Method, resourceName)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return &authorization.ResourceAttributes{
-		Namespace:   namespace,
+	r.Spec.ResourceAttributes = &authv1.ResourceAttributes{
+		Namespace:   pathSplit[5],
 		Verb:        verb,
-		Group:       group,
-		Version:     version,
+		Group:       pathSplit[2],
+		Version:     pathSplit[3],
 		Resource:    resource,
-		Subresource: subresource,
+		Subresource: pathSplit[8],
 		Name:        resourceName,
-	}, nil
+	}
+
+	return nil
+}
+
+func addNonNamespacedAttributes(req *restful.Request, r *authv1.SubjectAccessReview, pathSplit []string) error {
+	// URL example
+	// /apis/subresources.kubevirt.io/v1alpha3/expand-spec
+	if len(pathSplit) != 5 {
+		return fmt.Errorf("unknown api endpoint %s", req.Request.URL.Path)
+	}
+
+	resource := pathSplit[4]
+	if resource != "expand-spec" {
+		return fmt.Errorf("unknown resource type %s", resource)
+	}
+
+	verb, err := mapHttpVerbToRbacVerb(req.Request.Method, "")
+	if err != nil {
+		return err
+	}
+
+	r.Spec.ResourceAttributes = &authv1.ResourceAttributes{
```

Comment:
Can you add a comment here explaining that `expand-spec` is treated as a resource name by k8s, even though there is no CRD that defines it?

Reply:
Good point!